provider/datagen/src/source.rs

```diff
@@ -132,44 +147,93 @@ impl Debug for TomlCache {
 }
 
 impl TomlCache {
-    pub fn new(root: PathBuf) -> Self {
+    pub fn new(root: AbstractFs) -> Self {
         Self {
             root,
             cache: Arc::new(FrozenMap::new()),
         }
     }
 
-    pub fn read_and_parse_toml<S, P: AsRef<Path>>(&self, path: P) -> Result<&S, DataError>
+    pub fn read_and_parse_toml<S>(&self, path: &str) -> Result<&S, DataError>
     where
         for<'de> S: serde::Deserialize<'de> + 'static + Send + Sync,
     {
-        let path = self.root.join(path);
-
-        if self.cache.get(&path).is_none() {
-            log::trace!("Reading: {:?}", &path);
-            let file = std::fs::read_to_string(&path)
-                .map_err(|e| DataError::from(e).with_path_context(&path))?;
-            let file: S = toml::from_str(&file)
-                .map_err(|e| crate::error::data_error_from_toml(e).with_path_context(&path))?;
-            self.cache.insert(path.clone(), Box::new(file));
+        if self.cache.get(path).is_none() {
+            let file = self.root.open(&path)?;
+            let file: S = toml::from_slice(&file)
+                .map_err(|e| crate::error::data_error_from_toml(e).with_display_context(&path))?;
+            self.cache.insert(path.to_string(), Box::new(file));
         }
         self.cache
-            .get(&path)
+            .get(path)
             .unwrap()
             .downcast_ref::<S>()
             .ok_or_else(|| DataError::custom("Uprops TOML error").with_type_context::<S>())
     }
 
     #[cfg_attr(not(feature = "experimental"), allow(dead_code))]
     pub fn list(&self) -> Result<impl Iterator<Item = PathBuf>, DataError> {
-        let mut result = vec![];
-        for entry in std::fs::read_dir(&self.root)
-            .map_err(|e| DataError::from(e).with_path_context(&self.root))?
-        {
-            let entry = entry.map_err(|e| DataError::from(e).with_path_context(&self.root))?;
-            let path = entry.path();
-            result.push(path);
+        self.root.list("").map(Vec::into_iter)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum AbstractFs {
+    Fs(PathBuf),
+    // The PB points to a valid ZIP file, the String is a prefix inside the ZIP
+    Zip(PathBuf, String),
```

Comment:
thought: should we store the ZipArchive instead?

Reply:
The problem is that reading from the `ZipArchive` requires `&mut`. The next best thing would be to only create it once and then cheaply clone to get `&mut`s, but that's also not possible because `ZipArchive` is only `Clone` if the underlying `std::io::Read` is ( `File` isn't, so we'd be cloning `Vec<u8>`s).

Reply:
Did you try using a mutex to protect reads to the file?

(I think what you did is probably fine though)

Reply:
The file system supports parallel reads, if we put this behind a mutex we restrict ourselves.